src/main/java/org/nuxeo/audit/storage/impl/SQLAuditStorage.java

```diff
@@ -0,0 +1,196 @@
+
+	protected static final Log log = LogFactory.getLog(SQLAuditStorage.class);
+
+	protected static final String DIRECTORY_NAME = "sqlAuditDirectory";
+
+	protected static final String ID_COLUMN = "id";
+
+	protected static final String JSON_COLUMN = "entry";
+
+	protected CursorService<Iterator<LogEntry>, LogEntry> cursorService = new CursorService<>();
+
+	protected SQLDirectory sqlDirectory = (SQLDirectory) Framework.getService(DirectoryService.class)
+			.getDirectory(DIRECTORY_NAME);
+
+	/**
+	 * Insert entries as Json in the SQL Audit directory.
+	 */
+	@Override
+	public void append(List<String> jsonEntries) {
+		Session session = sqlDirectory.getSession();
+		for (String jsonEntry : jsonEntries) {
+			Map<String, Object> jsonMap = new HashMap<String, Object>();
+			jsonMap.put(JSON_COLUMN, jsonEntry);
+			session.createEntry(jsonMap);
+		}
+		session.close();
+	}
+
+	/**
+	 * Scroll log entries in the SQL Audit directory, given a scroll Id.
+	 */
+	@Override
+	public ScrollResult scroll(String scrollId) {
+		return cursorService.scroll(scrollId, true, logEntry -> String.valueOf(logEntry.getId()));
+	}
+
+	/**
+	 * Scroll log entries in the SQL Audit directory, given an audit query
+	 * builder.
+	 */
+	@Override
+	public ScrollResult scroll(AuditQueryBuilder queryBuilder, int batchSize, int keepAlive) {
+		cursorService.checkForTimedOutScroll();
+		List<LogEntry> logEntries = queryLogs(queryBuilder);
+		String scrollId = cursorService.registerCursor(logEntries.iterator(), batchSize, keepAlive);
+		return scroll(scrollId);
+	}
+
+	/**
+	 * Query log entries in the SQL Audit directory, given an audit query
+	 * builder. Does not support literals other than StringLiteral: see
+	 * {@link SQLSession.query(Map<String, Serializable>, Set<String>,
+	 * Map<String, String>, boolean, int, int)};
+	 */
+	protected List<LogEntry> queryLogs(AuditQueryBuilder queryBuilder) {
+		List<LogEntry> logEntries = new ArrayList<LogEntry>();
+
+		// Get the predicates filter map from the query builder.
+		Map<String, Serializable> filter = new HashMap<String, Serializable>();
+		MultiExpression predicate = (MultiExpression) queryBuilder.predicate();
+		@SuppressWarnings("unchecked")
+		List<Predicate> predicateList = (List<Predicate>) ((List<?>) predicate.values);
+		for (Predicate p : predicateList) {
+			String rvalue;
+			if (p.rvalue instanceof StringLiteral) {
+				rvalue = ((StringLiteral) p.rvalue).asString();
+			} else {
+				rvalue = p.rvalue.toString();
+				log.warn(String.format(
+						"Scrolling audit logs with a query builder containing non-string literals is not supported: %s.",
+						rvalue));
+			}
+			filter.put(p.lvalue.toString(), rvalue);
+		}
+
+		// Get the orderBy map from the query builder.
+		Map<String, String> orderBy = new HashMap<String, String>();
+		ArrayList<OrderByExpr> orders = queryBuilder.orders();
+		orders.stream().forEach(order -> orderBy.put(order.reference.name, order.isDescending ? "DESC" : "ASC"));
```

Comment:
Can be replaced with `Map<String, String> orderBy = orders.stream().collect(Collectors.toMap(o -> o.reference.name, o -> o.isDescending ? "DESC" : "ASC"));`

Reply:
+1, and note that even in the initial version `orders.stream().forEach` could have been simplified to `orders.forEach`, as `List` has directly a `forEach` method without going through streams.